File: Extensions/Tessa.Extensions.Default.Client/UI/TaskHistory/MakeViewTaskHistoryUIExtension.cs
#nullable enable
using System;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Input;
using Tessa.Cards;
using Tessa.Cards.TypeSettings;
using Tessa.Extensions.Default.Shared.Cards;
using Tessa.Extensions.Platform.Client.UI;
using Tessa.Platform;
using Tessa.Platform.Collections;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Cards.Tasks;
using Tessa.UI.Controls;
using Tessa.UI.Menu;
using Tessa.UI.Views.Content;

namespace Tessa.Extensions.Default.Client.UI.TaskHistory
{
    /// <summary>
    /// Реализация расширения типа карточки для добавления представлению функционала Истории заданий
    /// Добавляет стандартное контексное меню, тултип, обработку нажатий и открытие окна с деталями.
    /// Вторая часть расширения типа в WfTaskHistoryViewUIExtension
    /// </summary>
    public sealed class MakeViewTaskHistoryUIExtension(ICardDialogManager dialogManager) : CardUIExtension
    {
        #region Fields

        private readonly ICardDialogManager dialogManager = NotNullOrThrow(dialogManager);

        #endregion

        #region Type extension methods

        private static Task ExecuteInitializingAsync(ITypeExtensionContext typeContext)
        {
            var context = (ICardUIExtensionContext) typeContext.ExternalContext!;
            var settings = typeContext.Settings;
            var viewControlAlias = settings?.TryGet<string>(CardTypeExtensionSettings.ViewControlAlias);
            if (string.IsNullOrEmpty(viewControlAlias))
            {
                return Task.CompletedTask;
            }

            context.Model.ControlInitializers.Add((control, m, r, ct) =>
            {
                if (control is CardViewControlViewModel viewControl)
                {
                    if (viewControl.Name == viewControlAlias && viewControl.DataProvider is { } controlDataProvider)
                    {
                        var dataProvider = new TaskHistoryViewDataProvider(controlDataProvider);
                        viewControl.Refreshing += (s, e) => dataProvider.ResetCache();
                        viewControl.DataProvider = dataProvider;
                    }
                }

                return ValueTask.CompletedTask;
            });

            return Task.CompletedTask;
        }

        private Task ExecuteInitializedAsync(ITypeExtensionContext typeContext)
        {
            var context = (ICardUIExtensionContext) typeContext.ExternalContext!;

            var settings = typeContext.Settings;
            var viewControlAlias = settings?.TryGet<string>(CardTypeExtensionSettings.ViewControlAlias);
            if (string.IsNullOrEmpty(viewControlAlias))
            {
                return Task.CompletedTask;
            }

            if (context.Model.Controls.TryGet<CardViewControlViewModel>(viewControlAlias) is not { } taskHistoryView)
            {
                context.ValidationResult.AddException(this,
                    new ArgumentException($"Control ViewModel with Name='{viewControlAlias}' not found.", nameof(viewControlAlias)));
                return Task.CompletedTask;
            }

            var quickSearchIndex = taskHistoryView.TopItems.Items.IndexOf(i => i is QuickSearchViewModel);
            if (quickSearchIndex != -1)
            {
                taskHistoryView.TopItems.Items.RemoveAt(quickSearchIndex);
            }

            var clientQuickSearch = new ClientQuickSearchViewModel(taskHistoryView);
            taskHistoryView.TopItems.Items.Insert(quickSearchIndex != -1 ? quickSearchIndex : 0, clientQuickSearch);

            var expandAllButton = new ExpandAllButtonViewModel { ViewModel = taskHistoryView };
            taskHistoryView.TopItems.Add(expandAllButton);

            var collapseGroups = settings?.TryGet<bool>(CardTypeExtensionSettings.CollapseGroups) ?? false;

            var leftRowColumns = Array.Empty<string>();
            if (!string.IsNullOrWhiteSpace(settings?.TryGet<string>(CardTypeExtensionSettings.LeftRowColumns)))
            {
                leftRowColumns = settings.TryGet<string>(CardTypeExtensionSettings.LeftRowColumns)?.Split(' ');
            }

            var rightRowColumns = Array.Empty<string>();
            if (!string.IsNullOrWhiteSpace(settings?.TryGet<string>(CardTypeExtensionSettings.RightRowColumns)))
            {
                rightRowColumns = settings.TryGet<string>(CardTypeExtensionSettings.RightRowColumns)?.Split(' ');
            }

            var bottomColumns = Array.Empty<string>();
            if (!string.IsNullOrWhiteSpace(settings?.TryGet<string>(CardTypeExtensionSettings.BottomColumns)))
            {
                bottomColumns = settings.TryGet<string>(CardTypeExtensionSettings.BottomColumns)?.Split(' ');
            }

            var taskHistoryViewHelper = new TaskHistoryViewHelper(taskHistoryView, leftRowColumns, rightRowColumns, bottomColumns, collapseGroups);
            taskHistoryView.CreateRowFunc = taskHistoryViewHelper.CreateRow;

            taskHistoryView.RowContextMenuGenerators.Add(ctx =>
            {
                ctx.MenuActions.AddRange(
                    new MenuAction(
                        TaskHistoryMenuActionNames.Copy,
                        "$UI_Common_Copy",
                        Icon.Empty,
                        new DelegateCommand(p => CopyToClipboard(ctx.RowViewModel)),
                        "Ctrl+C"),
                    new MenuAction(
                        TaskHistoryMenuActionNames.ShowDetails,
                        "$UI_Cards_TaskHistory_ShowDetails",
                        Icon.Empty,
                        new DelegateCommand(p => { this.dialogManager.ShowTaskHistoryItem(taskHistoryViewHelper.CreateDetailsDialog(ctx.RowViewModel)); })),
                    new MenuAction(
                        TaskHistoryMenuActionNames.CollapseGroups,
                        "$UI_Cards_TaskHistory_ExpandGroups",
                        Icon.Empty,
                        new DelegateCommand(p => { ExpandGroups(ctx.RowViewModel); }),
                        isCollapsed: ctx.RowViewModel.Items.Count == 0),
                    new MenuAction(
                        TaskHistoryMenuActionNames.CollapseGroups,
                        "$UI_Cards_TaskHistory_CollapseGroups",
                        Icon.Empty,
                        new DelegateCommand(p => { CollapseGroups(ctx.RowViewModel); }),
                        isCollapsed: ctx.RowViewModel.Items.Count == 0)
                );

                return new ValueTask();
            });

            taskHistoryView.MiddleClickCommand = new DelegateCommand(o =>
            {
                var clickInfo = (IViewClickInfo) o;
                var row = (ViewControlRowViewModel) clickInfo.Row;
                if (row != null)
                {
                    this.dialogManager.ShowTaskHistoryItem(taskHistoryViewHelper.CreateDetailsDialog(row));
                }
            });

            taskHistoryView.KeyDownHandlers.Add((item, data, e) =>
            {
                if (data is ViewControlRowViewModel row)
                {
                    if (e.Key == Key.C && Keyboard.Modifiers == ModifierKeys.Control)
                    {
                        e.Handled = true;
                        CopyToClipboard(row);
                    }
                }
            });

            return Task.CompletedTask;
        }

        #endregion

        #region Private methods

        private static void CollapseGroups(TableRowViewModel row)
        {
            foreach (var item in row.Items)
            {
                CollapseGroups(item);
            }

            row.IsExpanded = false;
        }


        private static void ExpandGroups(TableRowViewModel row)
        {
            foreach (var item in row.Items)
            {
                ExpandGroups(item);
            }

            row.IsExpanded = true;
        }


        private static void CopyToClipboard(ViewControlRowViewModel row)
        {
            var sb = StringBuilderHelper.Acquire(256);

            foreach (var column in row.GridViewModel.Columns?.Cast<TableColumnViewModel>().Where(x => x.Visibility) ?? [])
            {
                if (sb.Length > 0)
                {
                    sb.AppendLine();
                }

                sb
                    .Append(column.Header)
                    .Append(": ")
                    .Append(Localize(row.CellsByColumnName[column.ColumnName].Text));
            }

            ClipboardSafe.SetText(sb.ToStringAndRelease());
        }

        #endregion

        #region Base overrides

        public override async Task Initializing(ICardUIExtensionContext context)
        {
            var result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.MakeViewTaskHistory,
                    context.Card,
                    context.Model.CardMetadata,
                    ExecuteInitializingAsync,
                    context,
                    cancellationToken: context.CancellationToken);

            context.ValidationResult.Add(result);
        }

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var result = await CardHelper
                .ExecuteTypeExtensionsAsync(
                    DefaultCardTypeExtensionTypes.MakeViewTaskHistory,
                    context.Card,
                    context.Model.CardMetadata,
                    this.ExecuteInitializedAsync,
                    context,
                    cancellationToken: context.CancellationToken);

            context.ValidationResult.Add(result);
        }

        #endregion
    }
}

File: WebClient SDK/src/default/ui/taskHistory/makeViewTaskHistoryUIExtension.ts
import { TaskHistoryViewDataProvider as TaskHistoryViewDataProvider } from './taskHistoryViewDataProvider';
import { ViewControlFilterableQuickSearchViewModel } from 'tessa/ui/cards/controls/viewControl/contents/viewControlFilterableQuickSearchViewModel';
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { tryGetFromSettings, MenuAction, showViewModelDialog } from 'tessa/ui';
import {
  ViewControlBaseButtonViewModel,
  ViewControlBlockMenuContext,
  ViewControlMultiSelectButtonViewModel,
  ViewControlPagingViewModel,
  ViewControlRowMenuContext,
  ViewControlTableGridViewModel,
  ViewControlToolbarItem,
  ViewControlViewModel
} from 'tessa/ui/cards/controls';
import { ValidationResult } from 'tessa/platform/validation';
import {
  ITableBlockViewModel,
  ITableRowViewModel,
  MultiSelectButtonViewModel,
  QuickSearchViewModel
} from 'tessa/ui/views/content';
import { LocalizationManager } from 'tessa/localization';
import { getTaskHistoryTooltip, TaskHistoryItemInfo } from 'tessa/ui/cards/tasks';
import { ViewMetadataSealed } from 'tessa/views/metadata';
import { TaskHistoryDetailsDialog } from 'tessa/ui/cards/components/forms';
import { Visibility } from 'tessa/platform';
import { runInAction } from 'mobx';
import { extension, localize } from '@tessa/application';
import { CardHelper, ICardTypeExtensionContext, CardTypeExtensionTypes } from '@tessa/platform';
import { ClipboardHelper } from 'tessa/ui/clipboard';

@extension()
export class MakeViewTaskHistoryUIExtension extends CardUIExtension {
  private _disposes: Array<Function | null> = [];

  public async initializing(context: ICardUIExtensionContext): Promise<void> {
    const result = await CardHelper.executeTypeExtensions(
      CardTypeExtensionTypes.MakeViewTaskHistory,
      context.card,
      context.model.generalMetadata,
      this.executeInitializingAction,
      context
    );

    context.validationResult.add(result);
  }

  public async initialized(context: ICardUIExtensionContext): Promise<void> {
    const result = await CardHelper.executeTypeExtensions(
      CardTypeExtensionTypes.MakeViewTaskHistory,
      context.card,
      context.model.generalMetadata,
      this.executeInitializedAction,
      context
    );

    context.validationResult.add(result);
  }

  public finalized(): void {
    for (const dispose of this._disposes) {
      if (dispose) {
        dispose();
      }
    }
    this._disposes.length = 0;
  }

  executeInitializingAction = async (typeContext: ICardTypeExtensionContext): Promise<void> => {
    const context = typeContext.externalContext as ICardUIExtensionContext;
    const settings = typeContext.settings;
    const viewControlAlias = tryGetFromSettings<string>(settings, 'ViewControlAlias');
    if (!viewControlAlias) {
      return;
    }

    context.model.controlInitializers.push(async control => {
      if (control instanceof ViewControlViewModel) {
        if (control.name === viewControlAlias) {
          const defaultDataProvider = control.dataProvider;
          if (!defaultDataProvider) {
            throw new Error(
              `Control ViewModel with Name='${viewControlAlias}' has not default data provider.`
            );
          }
          const dataProvider = new TaskHistoryViewDataProvider(defaultDataProvider);
          control.dataProvider = dataProvider;
          control.multiSelect = false;
          this._disposes.push(control.onRefreshing.addWithDispose(() => dataProvider.resetCache()));
        }
      }
    });
  };

  executeInitializedAction = async (typeContext: ICardTypeExtensionContext): Promise<void> => {
    const context = typeContext.externalContext as ICardUIExtensionContext;
    const settings = typeContext.settings;
    const viewControlAlias = tryGetFromSettings<string>(settings, 'ViewControlAlias');
    if (!viewControlAlias) {
      return;
    }

    const taskHistoryView = context.model.controls.get(viewControlAlias) as ViewControlViewModel;
    if (!taskHistoryView) {
      context.validationResult.add(
        ValidationResult.fromText(`Control ViewModel with Name='${viewControlAlias}' not found.`)
      );
      return;
    }

    const viewMetadata = taskHistoryView.viewMetadata!;

    const collapseGroups = tryGetFromSettings<boolean>(settings, 'CollapseGroups', false);

    const leftRowColumns = tryGetFromSettings<string>(settings, 'LeftRowColumns', '').split(' ');
    const rightRowColumns = tryGetFromSettings<string>(settings, 'RightRowColumns', '').split(' ');
    const bottomColumns = tryGetFromSettings<string>(settings, 'BottomColumns', '').split(' ');

    if (taskHistoryView.table) {
      const table = taskHistoryView.table;

      const defaultRowAction = table.createRowAction;
      table.createRowAction = opt => {
        const row = defaultRowAction(opt);
        row.toolTip = getTaskHistoryTooltip(
          this.getTaskHistoryItemInfo(
            viewMetadata,
            leftRowColumns,
            rightRowColumns,
            bottomColumns,
            row.data
          )
        );
        row.onMouseDown = e => {
          if (e.button === 1) {
            e.preventDefault();
            showViewModelDialog(
              this.getTaskHistoryItemInfo(
                viewMetadata,
                leftRowColumns,
                rightRowColumns,
                bottomColumns,
                row.data
              ),
              TaskHistoryDetailsDialog
            );
          }
        };

        row.isToggled = !collapseGroups;

        return row;
      };

      table.rowContextMenuGenerators.push(ctx => {
        ctx.menuActions.push(
          MenuAction.create({
            type: 'normal',
            name: 'Copy',
            caption: '$UI_Common_Copy',
            action: () => this.copyToClipboard(ctx.row)
          }),
          MenuAction.create({
            type: 'normal',
            name: 'ShowDetails',
            caption: '$UI_Cards_TaskHistory_ShowDetails',
            action: () =>
              showViewModelDialog(
                this.getTaskHistoryItemInfo(
                  viewMetadata,
                  leftRowColumns,
                  rightRowColumns,
                  bottomColumns,
                  ctx.row.data
                ),
                TaskHistoryDetailsDialog
              )
          })
        );
        if (ctx.tableGrid.rows.length > 0) {
          ctx.menuActions.push(...this.createGroupMenu(ctx));
        }
      });

      table.blockContextMenuGenerators.push(ctx => {
        ctx.menuActions.push(...this.createGroupMenu(ctx));
      });

      // ctrl + c
      this._disposes.push(
        table.keyDown.addWithDispose(e => {
          const currentItem = e.control.rows.find(x => x.isSelected);
          if (!currentItem) {
            return;
          }

          const { event } = e;
          const code = event.keyCode || event.charCode;
          if (code === 67 && event.ctrlKey) {
            this.copyToClipboard(currentItem);
          }
        })
      );
    }

    const paging = taskHistoryView.bottomItems.find(
      x => x.content instanceof ViewControlPagingViewModel
    );
    if (paging) {
      (paging.content as ViewControlPagingViewModel).rowCountVisibility = Visibility.Collapsed;
    }

    const quickSearchIndex = taskHistoryView.bottomItems.findIndex(
      x => x.content instanceof QuickSearchViewModel
    );
    if (quickSearchIndex > -1) {
      taskHistoryView.bottomItems.splice(quickSearchIndex, 1);
    }

    const multiSelect = taskHistoryView.bottomItems.find(
      x => x.content instanceof MultiSelectButtonViewModel
    );
    if (multiSelect) {
      (multiSelect.content as ViewControlMultiSelectButtonViewModel).visibility =
        Visibility.Collapsed;
    }

    const quickSearch = new ViewControlFilterableQuickSearchViewModel(taskHistoryView);
    quickSearch.initialize();
    taskHistoryView.bottomItems.splice(
      0,
      0,
      new ViewControlToolbarItem(quickSearch, 'right', {
        align: 'end',
        stretch: true,
        isPermanent: true
      })
    );

    const expandAllButton = new ViewControlBaseButtonViewModel(taskHistoryView);
    expandAllButton.initialize();
    expandAllButton.icon = 'm-expand-all';
    expandAllButton.caption = localize('$UI_Common_ExpandAll');
    expandAllButton.captionPosition = 'after';
    expandAllButton.showCaption = true;
    expandAllButton.tooltip = localize('$UI_Controls_Views_ExpandAll');
    expandAllButton.theme = 'transparent';
    expandAllButton.type = 'small';
    expandAllButton.onClick = () => {
      const table = taskHistoryView.table;
      if (!table) {
        return;
      }
      runInAction(() => {
        for (const block of table.blocks) {
          block.isToggled = true;
        }
        for (const row of table.rows) {
          row.isToggled = true;
        }
      });
    };
    taskHistoryView.bottomItems.push(new ViewControlToolbarItem(expandAllButton, 'right'));

    const collapseAllButton = new ViewControlBaseButtonViewModel(taskHistoryView);
    collapseAllButton.initialize();
    collapseAllButton.icon = 'm-fold-all';
    collapseAllButton.caption = localize('$UI_Common_CollapseAll');
    collapseAllButton.captionPosition = 'after';
    collapseAllButton.showCaption = true;
    collapseAllButton.tooltip = localize('$UI_Controls_Views_CollapseAll');
    collapseAllButton.theme = 'transparent';
    collapseAllButton.type = 'small';
    collapseAllButton.onClick = () => {
      const table = taskHistoryView.table;
      if (!table) {
        return;
      }
      runInAction(() => {
        for (const block of table.blocks) {
          block.isToggled = false;
        }
        for (const row of table.rows) {
          row.isToggled = false;
        }
      });
    };
    taskHistoryView.bottomItems.push(new ViewControlToolbarItem(collapseAllButton, 'right'));
  };

  private getTaskHistoryItemInfo(
    metadata: ViewMetadataSealed,
    leftColumns: string[],
    rightColumns: string[],
    bottomColumns: string[],
    row: ReadonlyMap<string, any>
  ): TaskHistoryItemInfo {
    return {
      leftItems: leftColumns.map(column => ({
        caption: metadata.columns.get(column)!.caption ?? '',
        data: row.get(column)
      })),
      rightItems: rightColumns.map(column => ({
        caption: metadata.columns.get(column)!.caption ?? '',
        data: row.get(column)
      })),
      bottomItems: bottomColumns.map(column => ({
        caption: metadata.columns.get(column)!.caption ?? '',
        data: row.get(column)
      }))
    };
  }

  private copyToClipboard(row: ITableRowViewModel) {
    let sb = '';
    for (const column of row.grid.columns.filter(x => x.visibility)) {
      if (sb) {
        sb += '\n';
      }

      sb += `${LocalizationManager.instance.localize(column.header)}: ${
        row.getByName(column.columnName)!.convertedValue
      }`;
    }

    ClipboardHelper.copyToClipboard(sb);
  }

  private createGroupMenu(
    ctx: ViewControlRowMenuContext | ViewControlBlockMenuContext
  ): ReadonlyArray<MenuAction> {
    const block = (ctx as ViewControlBlockMenuContext).block;
    const row = (ctx as ViewControlRowMenuContext).row;
    const table = ctx.tableGrid;
    return [
      MenuAction.create({
        type: 'normal',
        name: 'ExpandGroups',
        caption: '$UI_Cards_TaskHistory_ExpandGroups',
        action: () => {
          runInAction(() => {
            if (block) {
              this.toggleBlock(table, block, true);
            }
            if (row) {
              this.toggleRow(table, row, true);
            }
          });
        }
      }),
      MenuAction.create({
        type: 'normal',
        name: 'CollapseGroups',
        caption: '$UI_Cards_TaskHistory_CollapseGroups',
        action: () => {
          runInAction(() => {
            if (block) {
              this.toggleBlock(table, block, false);
            }
            if (row) {
              this.toggleRow(table, row, false);
            }
          });
        }
      })
    ];
  }

  private toggleRow(
    table: ViewControlTableGridViewModel,
    row: ITableRowViewModel,
    isToggled: boolean
  ) {
    row.isToggled = isToggled;
    for (const childRow of table.rows.filter(x => x.parentRowId === row.id)) {
      this.toggleRow(table, childRow, isToggled);
    }
  }

  private toggleBlock(
    table: ViewControlTableGridViewModel,
    block: ITableBlockViewModel,
    isToggled: boolean
  ) {
    block.isToggled = isToggled;
    for (const childBlock of table.blocks.filter(x => x.parentBlockId === block.id)) {
      this.toggleBlock(table, childBlock, isToggled);
    }
    for (const childRow of table.rows.filter(x => x.blockId === block.id)) {
      this.toggleRow(table, childRow, isToggled);
    }
  }
}
